File: PS_10_DivideAndConquer/MergeSort.cpp
#include <iostream>
#include <vector> // Include for std::vector
using namespace std;

vector<int> arr = { 12, 11, 13, 5, 6, 7 };

void merge(int left, int right)
{

  	int mid = (left + right) / 2;
    // Create temp vectors
    vector<int> temp;


    int i = left, j = mid + 1;

    // Merge the temp vectors back into arr[left..right]
    while (i <= mid && j <= right) {
        if (arr[i] <= arr[j]) {
            temp.push_back(arr[i]);
            i++;
        }
        else {
            temp.push_back(arr[j]);
            j++;
        }
    }

    // Copy the remaining elements
    for (; i <= mid; i++) {
      temp.push_back(arr[i]);
    }

    for (; j <= right; j++) {
      temp.push_back(arr[j]);
    }

    for (int i = 0; i < temp.size(); i++) {
      arr[left + i] = temp[i];
    }
}

// begin is for left index and end is right index
// of the sub-array of arr to be sorted
void mergeSort(int left, int right)
{
    if (left >= right)
        return;

    int mid = left + (right - left) / 2;
    mergeSort(left, mid);
    mergeSort(mid + 1, right);
    merge(left, right);
}

// Function to print a vector
void printVector(const vector<int>& arr)
{
    for (const int i : arr)
        cout << i << " ";
    cout << endl;
}

// Driver code
int main()
{
    int arrSize = arr.size();

    cout << "Given vector is \n";
    printVector(arr);

    mergeSort(0, arrSize - 1);

    cout << "\nSorted vector is \n";
    printVector(arr);
    return 0;
}

File: PS_11_TwoPointers/MergeSort.java
package myUsacoSilverSolutions.PS_11_TwoPointers;

public class MergeSort {
    static int[] arr = {3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5, 8, 9, 7, 8, 9};

    public static void merge(int start, int mid, int end) {
        int[] left = new int[mid - start + 1];
        int[] right = new int[end - mid];

        for (int i = 0; i < left.length; i++) {
            left[i] = arr[start + i];
        }

        for (int j = 0; j < right.length; j++) {
            right[j] = arr[mid + 1 + j];
        }

        int i = 0, j = 0, k = start;
        while (i < left.length && j < right.length) {
            if (left[i] <= right[j]) {
                arr[k] = left[i];
                i++;
            } else {
                arr[k] = right[j];
                j++;
            }
            k++;
        }

        while (i < left.length) {
            arr[k] = left[i];
            i++;
            k++;
        }

        while (j < right.length) {
            arr[k] = right[j];
            j++;
            k++;
        }
    }
    
    public static void mergesort(int[] arr, int start, int end) {
        if (start < end) {
            int mid = (start + end) / 2;
            mergesort(arr, start, mid);
            mergesort(arr, mid + 1, end);
            merge(start, mid, end);
        }
    }

    public static void main(String[] args) {
        mergesort(arr, 0, arr.length - 1);
        for (int i: arr)
            System.out.print(i + " ");
    }
}
